### DISCUSSION TITLE
clarifications regarding setPollTimeout function.

### DISCUSSION THREAD

**Discussion by vasantteja** [spring-projects/spring-kafka#2362] [Q&A] [answered]
I was researching on how to implement long polling in kafka similar to Amazon SQS which has an attribute called as "ReceiveMessageWaitTimeSeconds". Upon research I found kafka exposes this kind of functionality with consumer.poll(time). I was checking if spring-kafka also has same kind of implementation under the hood or not. Luckily I stumbled upon a thread on stackoverflow and github which confirm my hunch that consumer.poll(time) is similar to setPollTimeout function introduced in spring-kafka 2.3 Here are the links(https://github.com/spring-projects/spring-kafka/issues/248 and https://stackoverflow.com/questions/48402355/spring-kafkalistener-execute-and-poll-records-after-certain-interval. Can you please confirm if my understanding is correct or not?

**Comment by garyrussell** [comment, marked as answer]
`setPollTimeout` was **not** introduced in 2.3, it has been there from day 1 and, yes, that is the property passed to `consumer.poll()`.

The new 2.3 property is `idleBetweenPolls` where the polling thread sleeps for that time between calls to `poll()`.

**Reply by vasantteja** [reply]
Sorry my bad, I was under the wrong assumption.So the parameter we pass to consumer.poll() is similar to the "ReceiveMessageWaitTimeSeconds" attribute we use in SQS for long polling. How different is idleBetweenPolls from setPollTimeout and is this parameter same as "ReceiveMessageWaitTimeSeconds" in SQS? I am trying to correlate between parameters to understand long polling in kafka.

**Reply by garyrussell** [reply]
I am not familiar with SQS.

It sounds like `pollTimeout` is equivalent to `ReceiveMessageWaitTimeSeconds`; but there are other factors.

Kafka also has `fetch.min.bytes` and `fetch.max.wait.ms` where the `poll` will suspend (up to `pollTimeout`) until that number of bytes is available, or the max wait time has been exceeded.

`pollTimeout` really only applies if there are no records available to fetch, allowing the calling code to take control at that interval, when there are no records.

We added `idleBetweenPolls` at user request - for example, when they only want to receive `max.poll.records` every, say 60 seconds. i.e. fetch 10 records then wait a minute before fetching the next 10.

**Reply by vasantteja** [reply]
cool if I configure pollTimeout to 10ms and fetch.min.bytes to 1 and fetch.max.wait.ms to 20. In case when the data size is not more than 1 then poll is suspended till 20 ms. Did I get this right?

**Reply by garyrussell** [reply]
I am not sure what the behavior is if the poll timeout is less than `fetch.max.wait.ms`; my guess is that the `pollTimeout` wins.

I would hardly say the 10ms is "long" polling.

**Reply by vasantteja** [reply]
Got it. idleBetweenPolls is the attribute which we are going to use if we want to slow down the consumption on our client(consumer) side. One benefit I think is if the thread sleeps between the consumer.poll() calls we will reducing the cpu cycles. Is my understanding correct or is there any other benefit which I missed?
